stdlib/seq.mli

```diff
@@ -13,97 +13,807 @@
+    [cmp], then [sorted_merge cmp xs ys] is the sorted sequence obtained by
+    merging the sequences [xs] and [ys].
+
+    For more details on comparison functions, see {!Array.sort}.
+
+    @since 4.14 *)
+
+val product : 'a t -> 'b t -> ('a * 'b) t
+(** [product xs ys] is the Cartesian product of the sequences [xs] and [ys].
+
+    For every element [x] of [xs] and for every element [y] of [ys],
+    the pair [(x, y)] appears once as an element of [product xs ys].
+
+    The order in which the pairs appear is unspecified.
+
+    The sequences [xs] and [ys] are not required to be finite.
+
+    The sequences [xs] and [ys] must be persistent.
+
+    @since 4.14 *)
+
+val product_with : ('a -> 'b -> 'c) -> 'a t -> 'b t -> 'c t
+(** The sequence [product xs ys] is the image through [f] of the Cartesian
+    product of the sequences [xs] and [ys].
+
+    For every element [x] of [xs] and for every element [y] of [ys],
+    the element [f x y] appears once as an element of [product_with f xs ys].
+
+    The order in which these elements appear is unspecified.
+
+    The sequences [xs] and [ys] are not required to be finite.
+
+    The sequences [xs] and [ys] must be persistent.
+
+    [product_with f xs ys] is equivalent to
+    [map (fun (x, y) -> f x y) (product xs ys)].
+
+    @since 4.14 *)
+
+val ap : ('a -> 'b) t -> 'a t -> 'b t
+(** [ap] is the application operator of the type constructor [_ t],
+    viewed as an applicative functor.
+
+    For every element [x] of [xs] and for every element [y] of [ys],
+    the element [x y] appears once as an element of [ap xs ys].
+
+    The sequences [xs] and [ys] must be persistent.
+
+    [ap xs ys] is equivalent to [product_with (@@) xs ys].
+
+    @since 4.14 *)
+
+(** {1 Splitting a sequence into two sequences} *)
+
+val unzip : ('a * 'b) t -> 'a t * 'b t
+(** [unzip] transforms a sequence of pairs into a pair of sequences.
+
+    [unzip xs] is equivalent to [(map fst xs, map snd xs)].
+
+    Querying either of the sequences returned by [unzip xs]
+    causes [xs] to be queried.
+    Therefore, querying both of them
+    causes [xs] to be queried twice.
+    Thus, [xs] must be persistent and cheap.
+    If that is not the case, use [unzip (memoize xs)].
+
+    @since 4.14 *)
+
+val partition_either : ('a, 'b) Either.t t -> 'a t * 'b t
+(** [partition_either] transforms a sequence of fruit (either apples or oranges)
+    into a pair of a sequence of apples and a sequence of oranges.
+
+    [partition_either xs] is equivalent to
+    [(filter_map Either.find_left xs, filter_map Either.find_right xs)].
+
+    Querying either of the sequences returned by [partition_either xs]
+    causes [xs] to be queried.
+    Therefore, querying both of them
+    causes [xs] to be queried twice.
+    Thus, [xs] must be persistent and cheap.
+    If that is not the case, use [partition_either (memoize xs)].
+
+    @since 4.14 *)
+
+val partition : ('a -> bool) -> 'a t -> 'a t * 'a t
```

Comment:
4.12 also added:

```ocaml
module List : sig
  val partition_map : ('a -> ('b, 'c) Either.t) -> 'a list -> 'b list * 'c list
end
```

which could perhaps be considered for addition here for consistency, even if the implementation is analogous to `partition_either (map f x)`.

Reply:
This was discussed (in the affirmative) in https://github.com/ocaml/ocaml/pull/10583#discussion_r698504636